drivers/sensor/amd_sb_tsi/sb_tsi_emul.c

```diff
@@ -0,0 +1,157 @@
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/emul_sensor.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/logging/log.h>
+#include "sb_tsi.h"
+
+LOG_MODULE_DECLARE(AMD_SB_TSI, CONFIG_SENSOR_LOG_LEVEL);
+
+#define NUM_REGS 128
+
+struct sb_tsi_emul_data {
+	uint8_t reg[NUM_REGS];
+};
+
+struct sb_tsi_emul_cfg {
+};
+
+static void sb_tsi_emul_set_reg(const struct emul *target, uint8_t reg, uint8_t val)
+{
+	struct sb_tsi_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg < NUM_REGS);
+	data->reg[reg] = val;
+}
+
+static uint8_t sb_tsi_emul_get_reg(const struct emul *target, uint8_t reg)
+{
+	struct sb_tsi_emul_data *data = target->data;
+
+	__ASSERT_NO_MSG(reg < NUM_REGS);
+	return data->reg[reg];
+}
+
+static void sb_tsi_emul_reset(const struct emul *target)
+{
+	struct sb_tsi_emul_data *data = target->data;
+
+	memset(data->reg, 0, NUM_REGS);
+}
+
+static int sb_tsi_emul_transfer_i2c(const struct emul *target, struct i2c_msg *msgs,
+				    int num_msgs, int addr)
+{
+	/* Largely copied from emul_bmi160.c */
+	unsigned int val;
+	int reg;
+
+	__ASSERT_NO_MSG(msgs && num_msgs);
+
+	i2c_dump_msgs_rw("emul", msgs, num_msgs, addr, false);
+	switch (num_msgs) {
+	case 2:
+		if (msgs->flags & I2C_MSG_READ) {
+			LOG_ERR("Unexpected read");
+			return -EIO;
+		}
+		if (msgs->len != 1) {
+			LOG_ERR("Unexpected msg0 length %d", msgs->len);
+			return -EIO;
+		}
+		reg = msgs->buf[0];
+
+		/* Now process the 'read' part of the message */
+		msgs++;
+		if (msgs->flags & I2C_MSG_READ) {
+			switch (msgs->len) {
+			case 1:
+				val = sb_tsi_emul_get_reg(target, reg);
+				msgs->buf[0] = val;
+				break;
+			default:
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+				return -EIO;
+			}
+		} else {
+			if (msgs->len != 1) {
+				LOG_ERR("Unexpected msg1 length %d", msgs->len);
+			}
+			sb_tsi_emul_set_reg(target, reg, msgs->buf[0]);
+		}
+		break;
+	default:
+		LOG_ERR("Invalid number of messages: %d", num_msgs);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sb_tsi_emul_init(const struct emul *target, const struct device *parent)
+{
+	sb_tsi_emul_reset(target);
+	return 0;
+}
+
+static int sb_tsi_emul_set_channel(const struct emul *target, enum sensor_channel chan,
+				   q31_t value, int8_t shift)
+{
+	struct sb_tsi_emul_data *data = target->data;
+	int64_t scaled_value;
+	int temp_eighths;
+
+	if (chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+	scaled_value = (int64_t)value << shift;
+	temp_eighths = scaled_value >> 28;
+	temp_eighths = CLAMP(temp_eighths, 0, 0x7ff);
+
+	data->reg[SB_TSI_TEMP_INT] = temp_eighths >> 3;
+	data->reg[SB_TSI_TEMP_DEC] = (temp_eighths & 0x7) << 5;
+
+	return 0;
+}
+
+static int sb_tsi_emul_get_sample_range(const struct emul *target, enum sensor_channel chan,
+					q31_t *lower, q31_t *upper, q31_t *epsilon, int8_t *shift)
+{
+	if (chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+	*lower = 0;
```

Comment:
It looks like you're using these fields to pass around a raw register representation of the temperature rather than degrees Celsius in fixed-point format, which is the encoding the new sensors API as well as the backend emul API uses in place of the traditional `struct sensor_val`.

Based on what I could find this sensor runs a range of 0.125 C to 255.875 C. This will fit within the range of shift value 8 (Q9.23 in DSP notation), which is `-256` to `255.9999998807907`. The sensor's lower bound (0.125 C), for example, would be encoded as `0.125 * 2^31 >> 8 = 0x100000` (but I would just write the full expression instead of pasting the hex value for readability's sake: `*lower = (int64_t)(0.125 * ((int64_t)INT32_MAX + 1)) >> *shift;`). In this case, epsilon would be the same value because we go up in increments of 0.125 C. Epsilon is used as a tolerance when the generic test compares actual to expected to allow for rounding / precision errors. The value of `*upper` would be `0x7ff00000` (same math as above)

Let me know if you have any other questions, this can be hard to wrap one's head around at first. I found https://chummersone.github.io/qformat.html very useful. (For this case you'd plug in 32 as word size and 23 as the number of fractional bits)

Reply:
Thank you very much for the explanation. I applied the changes, except for the lower bound being 0C, not 0.125C